src/runtime-rs/crates/resource/src/cgroups/mod.rs

```diff
@@ -133,13 +216,90 @@ impl CgroupsResource {
 
         self.cgroup_manager
             .delete()
-            .context("delete cgroup manager")?;
+            .context("delete sandbox cgroup manager")?;
 
         if let Some(overhead) = self.overhead_cgroup_manager.as_ref() {
-            for cg_pid in overhead.tasks() {
-                overhead.remove_task(cg_pid)?;
+            for cg_pid in overhead.procs() {
+                overhead.remove_task_by_tgid(cg_pid)?;
+            }
+            overhead
+                .delete()
+                .context("delete overhead cgroup manager")?;
+        }
+
+        Ok(())
+    }
+
+    async fn delete_cgroup_v2(&self) -> Result<()> {
+        if self.cgroup_config.is_threaded_mode {
+            // Load parent cgroup manager.
+            let parent_cgroup_manager = self.cgroup_manager.parent_control_group();
+
+            // Move threads to parent sandbox cgroup manager from cgroup manager that thread mode.
+            for cg_tid in self.cgroup_manager.tasks() {
+                // For now, we can't guarantee that the thread in cgroup_manager does still
+                // exist. Once it exit, we should ignor that error returned by remove_task
+                // to let it go.
+                if let Err(error) = self.cgroup_manager.move_task_to_parent(cg_tid) {
+                    match error.source() {
+                        Some(err) => match err.downcast_ref::<io::Error>() {
+                            Some(e) => {
+                                if e.raw_os_error() != Some(OS_ERROR_NO_SUCH_PROCESS) {
+                                    return Err(error.into());
+                                }
+                            }
+                            None => return Err(error.into()),
+                        },
+                        None => return Err(error.into()),
+                    }
+                }
+            }
+
+            // Move threads to parent overhead cgroup manager from cgroup manager that thread mode.
+            if let Some(overhead) = self.overhead_cgroup_manager.as_ref() {
+                for cg_pid in overhead.tasks() {
+                    overhead.move_task_to_parent(cg_pid)?;
+                }
+            }
+
+            // Move procs to root cgroup manager from cgroup manager.
+            for cg_pid in parent_cgroup_manager.procs() {
+                parent_cgroup_manager.remove_task_by_tgid(cg_pid)?;
+            }
+
+            // Delete sandbox cgroup manager.
+            self.cgroup_manager
+                .delete()
+                .context("delete cgroup manager")?;
+
+            // Delete overhead cgroup manager.
+            if let Some(overhead) = self.overhead_cgroup_manager.as_ref() {
```

Comment:
Can line 276 and 259 be combined into one `if let ...`?

Reply:
merged together